unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs

```diff
@@ -0,0 +1,191 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class WalkerAgent : Agent
+{
+    [Header("Specific to Walker")] 
+    public Vector3 goalDirection;
+    public Transform hips;
+    public Transform chest;
+    public Transform spine;
+    public Transform head;
+    public Transform thighL;
+    public Transform shinL;
+    public Transform footL;
+    public Transform thighR;
+    public Transform shinR;
+    public Transform footR;
+    public Transform armL;
+    public Transform forearmL;
+    public Transform handL;
+    public Transform armR;
+    public Transform forearmR;
+    public Transform handR;
+    public Dictionary<Transform, BodyPart> bodyParts = new Dictionary<Transform, BodyPart>();
+
+    /// <summary>
+    /// Used to store relevant information for acting and learning for each body part in agent.
+    /// </summary>
+    [System.Serializable]
+    public class BodyPart
+    {
+        public ConfigurableJoint joint;
+        public Rigidbody rb;
+        public Vector3 startingPos;
+        public Quaternion startingRot;
+        public GroundContact groundContact;
+    }
+
+    /// <summary>
+    /// Create BodyPart object and add it to dictionary.
+    /// </summary>
+    public void SetupBodyPart(Transform t)
+    {
+        BodyPart bp = new BodyPart
+        {
+            rb = t.GetComponent<Rigidbody>(),
+            joint = t.GetComponent<ConfigurableJoint>(),
+            startingPos = t.position,
+            startingRot = t.rotation
+        };
+        bodyParts.Add(t, bp);
+        bp.groundContact = t.GetComponent<GroundContact>();
+    }
+
+    public override void InitializeAgent()
+    {
+        SetupBodyPart(hips);
+        SetupBodyPart(chest);
+        SetupBodyPart(spine);
+        SetupBodyPart(head);
+        SetupBodyPart(thighL);
+        SetupBodyPart(shinL);
+        SetupBodyPart(footL);
+        SetupBodyPart(thighR);
+        SetupBodyPart(shinR);
+        SetupBodyPart(footR);
+        SetupBodyPart(armL);
+        SetupBodyPart(forearmL);
+        SetupBodyPart(handL);
+        SetupBodyPart(armR);
+        SetupBodyPart(forearmR);
+        SetupBodyPart(handR);
+    }
+
+    /// <summary>
+    /// Obtains joint rotation (in Quaternion) from joint. 
+    /// </summary>
+    public static Quaternion GetJointRotation(ConfigurableJoint joint)
+    {
+        return (Quaternion.FromToRotation(joint.axis, joint.connectedBody.transform.rotation.eulerAngles));
+    }
+
+    /// <summary>
+    /// Add relevant information on each body part to observations.
+    /// </summary>
+    public void CollectObservationBodyPart(BodyPart bp)
+    {
+        var rb = bp.rb;
+        AddVectorObs(bp.groundContact.touchingGround ? 1 : 0); // Is this bp touching the ground
+        bp.groundContact.touchingGround = false;
```

Comment:
Why do we update `touchingGround`? A comment would be great.

Reply:
@Hunter-Unity ?